compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp

```diff
@@ -1072,10 +1072,15 @@ static void addLowerToLLVMGPUPasses(OpPassManager &modulePassManager,
       .addPass(createConvertComplexToStandardPass)
       // Convert BF16 operations to occur as F32.
       .addPass(createConvertBf16ArithToF32Pass)
-      .addPass(createConvertBf16ToUInt16BuffersPass)
-      // Convert math dialect elementry functions to polynomial form.
-      .addPass(createPolynomialApproximationPass)
-      .addPass(memref::createExpandOpsPass)
+      .addPass(createConvertBf16ToUInt16BuffersPass);
+
+  if (!forROCDL) {
+    // Convert math dialect elementry functions to polynomial form.
+    // On ROCM, convertToROCDL will expand these to device library calls.
+    funcPassManager.addPass(createPolynomialApproximationPass);
+  }
```

Comment:
I wanted to fix this too because exp2 is faster without polynomial approximation on rocm gpus. But I'm not sure if this way is very scalable. Can we instead create a LLVMGPUPolynomialApproximation pass? We can then enable/disable polynomial approximation based on what target we are using.

Reply:
> But I'm not sure if this way is very scalable

What do you mean by scalable?

I think that the polynomial approx should be always treated as a fallback for targets that don't have their own device lib exposing these functions. Any performant implementation will inherently be device-specific.

> We can then enable/disable polynomial approximation based on what target we are using.

I'd inverse this and say that it should be each target's responsibility to provide lowering for math functions, and targets should either do their own thing or use the generic polynomial approx pass. I think this should belong to passes later down the compilation pipeline, alongside convertToROCDL/NVVM/SPIRV/etc.

Reply:
> What do you mean by scalable?

I mean that each target has different requirements of what works and what doesn't work. Disabling polynomial approximation completley or enabling it for every target does not take this into account.

> I'd inverse this and say that it should be each target's responsibility to provide lowering for math functions, and targets should either do their own thing or use the generic polynomial approx pass. I think this should belong to passes later down the compilation pipeline, alongside convertToROCDL/NVVM/SPIRV/etc.

This is exactly what I'm saying too. Enable specific polynomial approximations based on target.

Reply:
But in this case we won't need an llvmgpu one because both rocm and nvvm handle it passes down the pass pipeline...